### ISSUE TITLE
False Positive: `rust.hyper.ssrf.reqwest-taint.reqwest-taint` triggers on macro-expanded std::fmt::format while ignoring equivalent `format!` macro

### ISSUE THREAD

**Issue by rikki322** [returntocorp/sgrep-rules#3746]
**Describe the bug**
Semgrep reports a false warning for the rule `rust.hyper.ssrf.reqwest-taint.reqwest-taint` when the code uses the expanded form of the `format!` macro (i.e., `std::fmt::format(format_args!(...))`).
The rule currently has logic to whitelist `format!` calls where the format string includes a hardcoded URL prefix (e.g., `format!("https://...", ...)`). However, this whitelist logic (specifically the `CONCAT` pattern in the rule) does not cover the underlying std::fmt::format implementation. So, the semantically equivalent code triggers a warning simply because the macro was expanded.

**To Reproduce**
The following input program contains two functions: Case 1 uses the macro and Case 2 uses the expanded form, both of them are semantics equivalent.
```
use std::collections::HashMap;
use std::net::SocketAddr;

use bytes::{Buf, Bytes};
use http_body_util::{BodyExt, Full};
use hyper::server::conn::http1;
use hyper::service::service_fn;
use hyper::{body::Incoming, header, Method, Request, Response, StatusCode};
use tokio::net::{TcpListener, TcpStream};

use reqwest::Client;

use url::form_urlencoded;

type GenericError = Box<dyn std::error::Error + Send + Sync>;
type Result<T> = std::result::Result<T, GenericError>;
type BoxBody = http_body_util::combinators::BoxBody<Bytes, hyper::Error>;

// Case 1: The rule correctly recognizes this pattern as safe or whitelisted.
async fn ok_api1(req: Request<Incoming>) -> Result<Response<BoxBody>> {
    let whole_body = req.collect().await?.aggregate();
    let mut data: serde_json::Value = serde_json::from_reader(whole_body.reader())?;
    let client = reqwest::Client::new();
    // ok: reqwest-taint
    let resp = client.get(format!("https://www.example.com/{}", data.hello))
        .send()
        .await?;

    println!("body = {:?}", resp);
    Ok("ok!")
}
// Case 2: Semantically identical to Case 1, but triggers the SSRF rule.
async fn ok_api2(req: Request<Incoming>) -> Result<Response<BoxBody>> {
    let whole_body = req.collect().await?.aggregate();
    let mut data: serde_json::Value = serde_json::from_reader(whole_body.reader())?;
    let client = reqwest::Client::new();
    // ok: reqwest-taint
    let resp = client.get(std::fmt::format(format_args!("https://www.example.com/{}", data.hello)))
        .send()
        .await?;

    println!("body = {:?}", resp);
    Ok("ok!")
}
```
Run Semgrep with the `rust.hyper.ssrf.reqwest-taint.reqwest-taint` rule

**Expected behavior**
Semgrep should not report `rust.hyper.ssrf.reqwest-taint.reqwest-taint` for the expanded code in Case 2. Since Case 1 is effectively whitelisted (likely because the rule identifies the `https://` prefix in the `format!` string), the expanded `std::fmt::format` version should also be recognized as a safe concatenation pattern.

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [x] P2: regular bug that should get fixed

**Environment**
- Semgrep version: 1.146.0
- Rule: `rust.hyper.ssrf.reqwest-taint.reqwest-taint`

**Comment by rikki322** [comment]
Update: additional affected rules
I also found that the same issue exists in the rule `rust.actix.ssrf.reqwest-taint.reqwest-taint` and rule `rust.rocket.ssrf.reqwest-taint.reqwest-taint`
## Code example
Run Semgrep with the `rust.actix.ssrf.reqwest-taint.reqwest-taint` rule
```
use actix_web::{post, web, HttpResponse, Result};
use reqwest::Client;
use serde::Deserialize;

#[derive(Deserialize)]struct Info {
    hello: String,
}

struct Collection<T>(T);
struct User;

// Case 1: The rule correctly recognizes this pattern as safe or whitelisted.
#[post("/ok_test1")]
async fn ok_test_case1(
  info: web::Path<Info>,
  col: web::Data<Collection<User>>,
) -> Result<String> {
  let client = reqwest::Client::new();
  // ok: reqwest-taint
  let resp = client.delete(format!("http://www.example.com/{}/123/{}", "name", info.hello))
    .send()
    .await?;

  println!("body = {:?}", resp);

  Ok(HttpResponse::Ok().body("Hello world!!!"))
}

// Case 2: Semantically identical to Case 1, but triggers the SSRF rule.
#[post("/ok_test2")]
async fn ok_test_case2(
  info: web::Path<Info>,
  col: web::Data<Collection<User>>,
) -> Result<String> {
  let client = reqwest::Client::new();
  // not ok: reqwest-taint
  let resp = client.delete(std::fmt::format(format_args!("http://www.example.com/{}/123/{}", "name", info.hello)))
    .send()
    .await?;

  println!("body = {:?}", resp);

  Ok(HttpResponse::Ok().body("Hello world!!!"))
}
```
Run Semgrep with the `rust.rocket.ssrf.reqwest-taint.reqwest-taint` rule
```
use rocket::{get, serde::json::Json};
use reqwest::Client;

// Mock structs for reproduction context
struct Db;
struct Connection<T>(T);
struct Post;

// Case 1: The rule correctly recognizes this pattern as safe or whitelisted.
#[get("/ok-test1")]
async fn ok_test_case1(mut db: Connection<Db>, id: String) -> Option<Json<Post>> {
    let client = reqwest::Client::new();
    // ok: reqwest-taint
    let resp = client.get(format!("https://www.example.com/{}", id))
        .send()
        .await;

    println!("body = {:?}", resp);
    Json(Post {})
}

// Case 2: Semantically identical to Case 1, but triggers the SSRF rule.
#[get("/ok-test2")]
async fn ok_test_case2(mut db: Connection<Db>, id: String) -> Option<Json<Post>> {
    let client = reqwest::Client::new();
    // not ok: reqwest-taint
    let resp = client.get(std::fmt::format(format_args!("https://www.example.com/{}", id)))
        .send()
        .await;

    println!("body = {:?}", resp);
    Json(Post {})
}
```
